### PR TITLE
Support service binding type 'npmrc'

### LINKED ISSUE #308: Configure `NPM_CONFIG_GLOBALCONFIG`
Implement https://github.com/paketo-buildpacks/nodejs/issues/396 in the NPM Install buildpack.

## Implementation Notes
The implementation should set the `NPM_CONFIG_GLOBALCONFIG` environment variable at build-time for use within the `npm install` command, and for subsequent buildpacks that rely on the node modules (such as the Node Run Script buildpack).

### PR DESCRIPTION
**Pull request by fg-j**
## Summary

Resolves #308 

## Use Cases

We made a few decisions in this implementation that we want to call out:
1. If more than one binding of type `npmrc` is provided, the build fails, since there's no way to set `NPM_CONFIG_GLOBALCONFIG` to multiple file locations.
2. If an `npmrc` service binding is provided, `NPM_CONFIG_GLOBALCONFIG` is set to its path as a **default** environment variable on the node modules layer. As a result, if a user manually sets `NPM_CONFIG_GLOBALCONFIG` at build time, the user-provided value gets priority over the value determined by the service binding resolver.

Feedback is welcome on how these choices impact the value of the feature.

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by fg-j** [comment]
Blocked on https://github.com/paketo-buildpacks/occam/pull/119

**Review by ryanmoran — COMMENTED** [review]
Should we try to avoid passing the `npmrc` filepath around through an environment variable? Instead, we could pass it explicitly to the places it is used, `Environment` and the `*BuildProcess` types. This would arguably be much more explicit and understandable in the code.

**Comment by fg-j** [comment]
When you say "avoid passing the npmrc filepath around through an environment variable", does that suggest changing `os.Setenv("NPM_CONFIG_GLOBALCONFIG")` , but still setting `nodeModulesLayer.BuildEnv` and  `buildProcess.environment`, @ryanmoran?

**Comment by ryanmoran** [comment]
I mean the former. I'd like to see us try to avoid using the environment of the local process as a storage mechanism for passing variables between objects. We can pass the strings around directly.

File: build_test.go
Comment (fg-j) [inline build_test.go:312]: These bindings can be simplified -- they don't need `Entries`

**Comment by fg-j** [comment]
The integration tests are currently failing during the `node-engine` build process due to a permissions error. I'm unable to reproduce this on my machine. See https://github.com/paketo-buildpacks/npm-install/runs/4731148661?check_suite_focus=true#step:5:369.
@paketo-buildpacks/nodejs-maintainers Can you take a look at what's going on here?

**Comment by ryanmoran** [comment]
@fg-j I believe you are running into this: https://stackoverflow.com/questions/61009190/docker-mount-permissions-on-mac-vs-linux. On macos, the host makes sure that the permissions line up with the user inside the container for ease of use while that is not how the bind mounting behavior works on Linux. I believe this is compounded by the way the binding directory is crated as `os.MkdirTemp` uses `0700` as the permissions for the directory it creates. Since the UID inside the container will not match the UID on the host and we have denied permissions for group and everyone, we see a permission denied error when we try to find files in that directory. Specifically, the service bindings code in `packit` is trying to differentiate between old-style bindings and new-style bindings by looking for a file called `metadata`. However, because the directory cannot be listed by the container user, we just get a permission denied error.

I think the answer is to `os.Chmod` the directory to something more permissive like `os.ModePerm`.

**Comment by fg-j** [comment]
@ryanmoran That makes sense. Is the reason we're only seeing this crop up now is because we previously haven't integration tested build-time service bindings?

For now, this PR is blocked on the fix in  `servicebindings`.

File: build.go
Comment (fg-j) [inline build.go:72]: Should use `servicebindings.ResolveOne()` instead of this logic.

Comment (fg-j) [reply build.go:72]: Unfortunately, `servicebindings.ResolveOne()` fails no binding of the provided type is found. We can't use it here since the `npmrc` binding is optional

Comment (noamichael) [reply build.go:72]: Could you do this?

```go
serviceBindingResolver := servicebindings.NewResolver()
npmrc, err := serviceBindingResolver.ResolveOne("npmrc", "", context.Platform.Path)

// err == nil ONLY if binding exists and there was no IO error
if err == nil {
	logger.Process("Using .npmrc binding")
	npmrcPath = filepath.Join(npmrc.Path, ".npmrc")
}
```

**Review by ryanmoran — CHANGES_REQUESTED** [review]

File: build.go
Comment (fg-j) [reply build.go:75]: 😂  brain glitch. good catch.

**Review by ryanmoran — APPROVED** [review]

**Pull request merged**
